feat(profile): Profile 관련 페이지 퍼블리싱

x0gux opened the pull request:
## 📋 관련 이슈

Closes #

## 📝 작업 내용

프로필 렌더링 및 수정을 퍼블리싱했습니다

## 📸 스크린샷 (UI 변경 시)

<!-- 변경된 UI가 있다면 스크린샷을 첨부해주세요 -->

## ✅ 체크리스트

- [ ] 관련 이슈 연결
- [ ] 셀프 코드 리뷰 완료
- [ ] 컨벤션 준수 확인
- [ ] 린트/빌드 통과 확인


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Complete profile experience: full profile page with header, editable info, picture upload, bio, organization, custom links, roadmap lists, add-roadmap modal, and contribution graph with streak/stats.

* **UI**
  * New building blocks: avatar, badge, scroll area, separator, and textarea for consistent presentation.

* **Tests**
  * Extensive unit tests covering profile components and interactions.

* **Chores**
  * Centralized exports and profile state scaffolding to support the feature set.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

x0gux committed:
wip: profile 파일 구조

x0gux committed:
wip: profile 이미지 완성

x0gux committed:
Merge remote-tracking branch 'origin' into feat/#15-profile-publishing

x0gux committed:
wip: profile 유저정보 인풋 작업

x0gux committed:
wip: profile 헤더 작업

x0gux committed:
wip: profile 렌더링

x0gux committed:
wip: profile 수정 작업

x0gux committed:
feat(profile): 프로필 페이지 퍼블리싱

justn-hyeok commented:
드디어 ㅜㅜㅜ

x0gux commented:
니가 보기에 굉장히 맘에 안들수있지만 ㅈㄴ 최선을 다했다

justn-hyeok commented:
파일수정 41개 뒤질래? @taehyun00

justn-hyeok commented on src/features/profile/utils/contribution-utils.tsx:
여기 주석은 불필요한 거 같아요

justn-hyeok commented on src/features/profile/components/templates/Profile.tsx:
```diff
@@ -0,0 +1,18 @@
+  return (
+    <div className="flex w-[950px] flex-col gap-12">
+      {ProfileHeader()}
+      <ProfileBio bio="임베디드 소프트웨어과 카사노비 김태현 여자 500,000,000명을 따먹다. 그냥 다 먹어버리고싶노 ㅋㅋㅋㅋㅋㅋㅋ 비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비누를먹어누를먹어비누를먹어비누를먹어비누를먹어비누를먹어비 ...
```
ㅇㅇ 여기 목 데이터 너무 에바임. 좀 얌전하게 고쳐주실 수 있을까요

justn-hyeok commented on src/features/profile/stores/profile-atoms.ts:
```diff
@@ -0,0 +1,6 @@
+import { atom } from 'jotai';
+
+export const profileModeAtom = atom<'show' | 'edit'>('show');
+export const profileBioAtom = atom('임베디드 소프트웨어과 카사노바 김태현');
```
이거도 기본값을 설정하던가 좀 더 적절한 목을 넣던가 하는게 좋을 거 같고

justn-hyeok commented on src/features/profile/components/templates/Profile.tsx:
```diff
@@ -0,0 +1,18 @@
+export default function Profile() {
+  return (
+    <div className="flex w-[950px] flex-col gap-12">
+      {ProfileHeader()}
```
```suggestion
      <ProfileHeader />
```

올바른 컴포넌트 호출 방식을 사용해주세요

justn-hyeok commented on src/features/profile/components/atoms/ProfilePicture.tsx:
```diff
@@ -0,0 +1,56 @@
+          >
+            <Pencil size={16} />
+          </Button>
+          <input id="profile-upload" type="file" className="hidden" accept="image/*" />
```
onChange 핸들러가 없어서 파일 선택해도 아무 일도 안 일어나요. 파일 업로드 핸들러 추가 필요할 거 같네요

justn-hyeok commented on src/features/profile/components/molecules/ProfileInfoForm.tsx:
```diff
@@ -0,0 +1,75 @@
+  };
+
+  const [names, setName] = useState(name);
+  const [emails, setEmail] = useState(email);
```
```suggestion
  const [name, setName] = useState(name);
  const [email, setemail] = useState(email);
```

justn-hyeok commented on src/features/profile/components/atoms/ProfileCustomLink.tsx:
```diff
@@ -0,0 +1,85 @@
+                const next = [...links];
+                next[index].name = e.target.value;
+                setLinks(next);
+              }}
```
```suggestion
              onChange={(e) => {
                const next = [...links];
                next[index] = { ...next[index], name: e.target.value };
                setLinks(next);
              }}
```
불변성 유지하도록 수정해주세요

justn-hyeok commented on src/features/profile/components/atoms/ProfileCustomLink.tsx:
```diff
@@ -0,0 +1,85 @@
+                  const next = [...links];
+                  next[index].url = e.target.value;
+                  setLinks(next);
+                }}
```
이거도 포함

justn-hyeok commented on src/features/profile/components/atoms/AddRoadmapModal.tsx:
```diff
@@ -0,0 +1,201 @@
+    );
+  }
+
+  // File Item
```
지금 이런 식의 불필요한 주석 정도는 제거해주세요.

justn-hyeok commented on src/features/profile/components/atoms/ContributionGraph.tsx:
```diff
@@ -0,0 +1,49 @@
+  return (
+    <div className="flex h-[137px] w-[552px] flex-col items-center justify-center rounded-md border">
+      <div className="w-[100%] p-2">
+        <p className="text-base font-bold text-slate-500">348일 연속 스트릭</p>
```
여기도 지금 왜 data를 받으면서 이렇게 목으로 받는건가요

justn-hyeok commented on src/features/profile/components/atoms/ProfileBio.tsx:
```diff
@@ -0,0 +1,80 @@
+  const mode = useAtomValue(profileModeAtom);
+  const [isExpanded, setIsExpanded] = useState(false);
+
+  const [bios, setbio] = useState(bio);
```
```suggestion
  const [bio, setBio] = useState(bio);
```

justn-hyeok commented on src/features/profile/components/atoms/ProfilePicture.tsx:
```diff
@@ -0,0 +1,56 @@
+    className="h-full w-full object-cover"
+    width={100}
+    height={100}
+  />
```
```suggestion
  <Image
    src={src}
    alt="profile picture"
    fill
    className="object-cover"
  />
```

<Image /> 똑바로 사용해주세요

justn-hyeok commented on src/features/profile/components/atoms/RoadmapCard.tsx:
```diff
@@ -0,0 +1,19 @@
+export function RoadmapCard({ title, author }: RoadmapCardProps) {
+  return (
+    <Card className="h-[200px] w-[304px] overflow-hidden rounded-xl border border-slate-200 bg-white shadow-none">
+      <div className="mt-[-30px] h-[190px] border-b border-slate-200 bg-slate-100" />
```
negative margin으로 위치 조정하는 것은 취약함. 정상적인 레이아웃으로 설계해주세요

justn-hyeok commented on src/features/profile/components/atoms/ProfileCustomLink.tsx:
```diff
@@ -0,0 +1,85 @@
+
+          {link.name && <p className="text-sm font-medium text-slate-900">{link.name}</p>}
+
+          <p className="flex-1 truncate text-sm text-slate-500 underline">{link.url}</p>
```
아니 이거 왜 p 태그임 ㅋㅋㅋㅋ 링크잖아요 a로 수정해주세요

justn-hyeok commented on src/features/profile/components/atoms/ProfileCustomLink.tsx:
생각해보면 우리 지금 링크 삭제 관련 로직이 없는듯.

justn-hyeok commented on src/features/profile/components/atoms/ContributionGraph.tsx:
```diff
@@ -0,0 +1,49 @@
+
+  const map = new Map(data.map((d) => [d.date, d.count]));
+
+  const weeks = chunkByWeek(padded.map((d) => (d ? { date: d, count: map.get(d) ?? 0 } : null)));
```
재계산이 불필요함. useMemo로 최적화합시다.

justn-hyeok commented on src/features/profile/components/atoms/ProfileBio.tsx:
```diff
@@ -0,0 +1,80 @@
+          }}
+          value={bios}
+          onChange={(e) => setbio(e.target.value)}
+        />
```
인라인 스타일 넣지마여

justn-hyeok reviewed (changes requested):
지금 사실 atomic design도 많이 어겼어요. 피드백 남겨드릴게요.

- [ ] auth는 지금 배럴 파일이 존재하는 반면에 profile은 없음.
- [ ] 현재 atoms/ 안에 너무 많은 파일이 존재하고, 대부분이 계층을 어김.
  - atoms/는 '더 이상 쪼갤 수 없는 최소 단위'
  - [ ] 맨 밑에 올바른 파일 위치를 제안해드릴게요. 수정 부탁드려요.
- [ ] molecules가 너무 얇고 이름이 불명확해요.
- [ ] 컴포넌트 export 방식 불일치
  - [ ] 전부 named export로 통일해주세요. 그게 트리 쉐이킹에 유리해요.
- [ ] 현재 파일명과 컴포넌트명 불일치하는게 존재해요. 확인 후 수정해주세요.
- [ ] utils 파일 확장자가 불일치해요. 
- [ ] schemas vs types 폴더 부재해요. ProfilePictureProps, ProfileBioProps 등이 각 파일에 분산돼있어요.


```
  atoms/
    ├── ProfilePicture.tsx        ✅ 유지
    ├── ProfileEditButton.tsx     ✅ 유지
    ├── RoadmapCard.tsx          ✅ 유지
    └── ProfileLinkAddButton.tsx ✅ 유지

  molecules/
  ├── ProfileBio.tsx           ⬅️ atoms에서 이동
  ├── ProfileCustomOrganization.tsx ⬅️ atoms에서 이동
  ├── ProfileInfoForm.tsx      ⬅️ ProfileInfomation 리네임 + 이동
  ├── ContributionGraph.tsx    ⬅️ atoms에서 이동
  ├── RoadmapList.tsx         ⬅️ atoms에서 이동
  ├── ProfileHeader.tsx        ✅ 유지
  ├── ProfileCustomBoxArea.tsx ✅ 유지
  └── ProfileStric.tsx        ✅ 유지

  organisms/
  ├── AddRoadmapModal.tsx     ⬅️ atoms에서 이동
  ├── MadeRoadmapList.tsx     ⬅️ atoms에서 이동
  ├── ProfileCustomLinks.tsx  ⬅️ ProfileCustomLink 리네임 + 이동
  ├── ProfileSecondBox.tsx    ✅ 유지
  └── ProfileThirdBox.tsx     ✅ 유지
```

justn-hyeok commented:
구조 자체는 auth 참고하면서 하시면 편할 거 같아요.

justn-hyeok commented on src/features/profile/components/atoms/ContributionGraph.tsx:
```diff
@@ -0,0 +1,49 @@
+
+  return (
+    <div className="flex h-[137px] w-[552px] flex-col items-center justify-center rounded-md border">
+      <div className="w-[100%] p-2">
```
w-full로 수정 ㄱㄱ

justn-hyeok commented on src/features/profile/components/atoms/ProfilePicture.tsx:
```diff
@@ -0,0 +1,56 @@
+const StyledImage = ({ src }: { src: string }) => (
+  <Image
+    src={src}
+    alt="profile picture"
```
```suggestion
    alt="{`${userName}'s profile picture`}"
```

이거 접근성 너무 구짐. 수정해주세요.

justn-hyeok commented on src/features/profile/components/molecules/ProfileStric.tsx:
```diff
@@ -0,0 +1,11 @@
+export default function ProfileStric() {
+  const data = GenerateMockContributions();
+  return (
+    <div>
```
이거 div 왜 있음 ??

justn-hyeok commented on src/features/profile/components/molecules/ProfileStric.tsx:
ProfileStreak로 바꿔주세요 ..

justn-hyeok reviewed (commented):
추가로 커밋 올라오고 코멘트 코드래빗 포함 다 해결하면 저 맨션해주세요 @taehyun00

x0gux committed:
refactor(profile): 코멘트 반영

justn-hyeok reviewed (changes requested):
일단 코드래빗 피드백을 다 반영해보죠. 저거 다 반영할만한듯??

justn-hyeok reviewed (changes requested):
## 🔍 코드 리뷰 - Profile Feature

auth 피처와 비교 분석한 결과입니다.

---

### 🔴 **치명적 이슈**

#### 1. `'use client'` 누락
다음 파일들에 클라이언트 디렉티브가 누락되었습니다:

| 파일 | 사용 훅 |
|------|--------|
| `ProfilePicture.tsx` | `useRef`, `useAtomValue` |
| `ProfileInfoForm.tsx` | `useState`, `useAtom` |

```diff
+ 'use client';
+
  import { useRef } from 'react';
```

#### 2. Tailwind `!` 문법 deprecated
`ProfileEditButton.tsx:19`
```typescript
// ❌ 현재
className={cn('... bg-indigo-950! ... text-white!', className)}

// ✅ 수정
className={cn('... bg-indigo-950 ... text-white', className)}
```
> Tailwind v4에서 `!` prefix 문법은 deprecated입니다. 제거하거나 `bg-indigo-950!` → `bg-indigo-950` 형태로 수정해주세요.

---

### 🟠 **주요 이슈**

#### 3. State 네이밍 불일치
`ProfileInfoForm.tsx:22-23`, `ProfileBio.tsx:17`
```typescript
// ❌ 현재 - 복수형 변수, 단수형 setter
const [names, setName] = useState(name);
const [emails, setEmail] = useState(email);
const [bios, setBios] = useState(bio);

// ✅ 권장
const [userName, setUserName] = useState(name);
const [userEmail, setUserEmail] = useState(email);
const [userBio, setUserBio] = useState(bio);
```

#### 4. 안전하지 않은 ID 생성
`ProfileCustomLinks.tsx:40`
```typescript
// ❌ 현재 - Math.random() 유니크 보장 안됨, substr() deprecated
id: Math.random().toString(36).substr(2, 9)

// ✅ 권장
id: crypto.randomUUID()
```

#### 5. `cn()` 대신 템플릿 리터럴 사용
`ProfileBio.tsx:36`
```typescript
// ❌ 현재
className={`${!isExpanded ? 'line-clamp-3' : ''} text-justify...`}

// ✅ auth 피처 패턴과 일치
className={cn('text-justify...', !isExpanded && 'line-clamp-3')}
```

---

### 📊 Auth vs Profile 패턴 비교

| 항목 | Auth | Profile | 상태 |
|------|------|---------|------|
| `'use client'` | 일관됨 | 2개 누락 | ❌ |
| Tailwind 문법 | `cn()` + 최신 | `!` deprecated | ❌ |
| State 네이밍 | 일관됨 | 불일치 | ❌ |

---

위 이슈들 수정 부탁드립니다! 🙏

justn-hyeok commented on src/features/profile/components/molecules/ProfileInfoForm.tsx:
```diff
@@ -0,0 +1,74 @@
+import { useState } from 'react';
```
## 🔴 `'use client'` 누락

`useState`와 `useAtom` 훅을 사용하고 있어서 클라이언트 컴포넌트로 선언해야 합니다.

```diff
+ 'use client';
+
  import { useState } from 'react';
```

justn-hyeok commented on src/features/profile/components/atoms/ProfileEditButton.tsx:
```diff
@@ -0,0 +1,25 @@
+    <Button
+      type="button"
+      variant="outline"
+      className={cn('border-border w-full bg-indigo-950! font-semibold text-white!', className)}
```
## 🟠 Tailwind `!` 문법 deprecated

Tailwind v4에서 `!` prefix 문법은 deprecated입니다.

```diff
- className={cn('... bg-indigo-950! ... text-white!', className)}
+ className={cn('... bg-indigo-950 ... text-white', className)}
```

`!important`가 필요하다면 다른 방식 검토 필요

justn-hyeok commented on src/features/profile/components/organisms/ProfileCustomLinks.tsx:
```diff
@@ -0,0 +1,110 @@
+  };
+
+  const handleAddLink = () => {
+    const next = [...links, { id: Math.random().toString(36).substr(2, 9), name: '', url: '' }];
```
## 🟠 안전하지 않은 ID 생성

`Math.random()`은 유니크 보장이 안 되고, `substr()`은 deprecated입니다.

```diff
- id: Math.random().toString(36).substr(2, 9)
+ id: crypto.randomUUID()
```

justn-hyeok commented on src/features/profile/components/molecules/ProfileBio.tsx:
```diff
@@ -0,0 +1,65 @@
+    <div className="flex flex-col gap-4">
+      <p className="text-sm font-semibold">자기소개</p>
+      <div className="flex flex-col gap-2 rounded-lg border border-slate-200 p-2">
+        <p className={`${!isExpanded ? 'line-clamp-3' : ''} text-justify text-sm text-slate-500`}>
```
## 🟡 `cn()` 사용 권장

auth 피처와 일관성을 위해 템플릿 리터럴 대신 `cn()` 유틸 사용을 권장합니다.

```diff
- className={`${!isExpanded ? 'line-clamp-3' : ''} text-justify...`}
+ className={cn('text-justify...', !isExpanded && 'line-clamp-3')}
```

justn-hyeok commented on src/features/profile/components/molecules/ProfileInfoForm.tsx:
```diff
@@ -0,0 +1,74 @@
+    setMode((prev) => (prev === 'show' ? 'edit' : 'show'));
+  };
+
+  const [names, setName] = useState(name);
```
## 🟡 State 네이밍 불일치

변수명과 setter 이름이 일치하지 않습니다.

```diff
- const [names, setName] = useState(name);
- const [emails, setEmail] = useState(email);
+ const [userName, setUserName] = useState(name);
+ const [userEmail, setUserEmail] = useState(email);
```

복수형 변수에 단수형 setter는 혼란을 줄 수 있어요.

justn-hyeok commented on src/features/profile/components/molecules/ProfileBio.tsx:
```diff
@@ -0,0 +1,65 @@
+  const mode = useAtomValue(profileModeAtom);
+  const [isExpanded, setIsExpanded] = useState(false);
+
+  const [bios, setBios] = useState(bio);
```
## 🟡 State 네이밍 불일치

단일 값인데 복수형 이름 사용 중입니다.

```diff
- const [bios, setBios] = useState(bio);
+ const [userBio, setUserBio] = useState(bio);
```

x0gux committed:
refactor(profile): 코멘트 반영

justn-hyeok committed:
refactor(profile): address PR #21 review feedback

- fix invalid Tailwind class and add follower/following props
- add aria-labels for accessibility
- extract RoadmapCardItem component
- reset modal state on close and fix folder filtering

justn-hyeok committed:
fix(profile): 추가적인 코드래빗 코드 리뷰 수정

justn-hyeok committed:
fix: tailwind className을 시멘틱 태그로 번경

justn-hyeok committed:
fix: figma에 맞는 스택으로 버튼 디자인 변경

justn-hyeok committed:
fix: shadcn 기반으로 컴포넌트 번경

justn-hyeok committed:
fix: 피그마에 맞춰서 바꿔

justn-hyeok committed:
wip: 퍼블리싱 수정, 피그마에 맞게 이어서해주세요 ㅠㅠ @taehyun00

x0gux committed:
refactor(profile): 코멘트 반영

x0gux committed:
refactor(profile): profilebio shadcn 변경

justn-hyeok commented:
<img width="1033" height="288" alt="image" src="https://github.com/user-attachments/assets/7cff92b5-e7fb-45bf-92f8-295ed79e4782" />
여기 어색함...





<img width="584" height="173" alt="스크린샷 2026-01-02 오전 12 11 04" src="https://github.com/user-attachments/assets/dcc705d0-cf0c-421e-b757-2e1ca75f7e06" />
간격 어색함...

<img width="924" height="735" alt="스크린샷 2026-01-02 오전 12 11 19" src="https://github.com/user-attachments/assets/8af07ff0-28cc-4f62-8580-d2078d87e02e" />
사진 들어가간 거 어색햐게 들어감. 피그마랑 대조하면서 ㄱㄱ

그리고 값 없을 때 뜰 텍스트 같은 거 적용만 하면 ㄹㅇ 될 듯

x0gux committed:
refactor(profile): profile 코멘트 반영

x0gux committed:
refactor(profile): github action test

justn-hyeok reviewed (approved):
고생햇다 ㅜㅜ
